include/dlaf/factorization/qr/t_factor_impl.h

```diff
@@ -0,0 +1,347 @@
+                          pika::unwrapping([](auto&& tile, cudaStream_t stream) {
+                            tile::internal::set0<T>(tile, stream);
+                            return std::move(tile);
+                          }),
+                          std::move(t));
+  }
+
+  static pika::future<matrix::Tile<T, Device::GPU>> gemvColumnT(
+      bool is_v0, pika::shared_future<matrix::Tile<const T, Device::GPU>> tile_vi,
+      pika::shared_future<common::internal::vector<T>>& taus,
+      pika::future<matrix::Tile<T, Device::GPU>>& tile_t) noexcept {
+    auto gemv_func = pika::unwrapping(
+        [is_v0](cublasHandle_t handle, const auto& tile_v, const auto& taus, auto&& tile_t) {
+          const SizeType k = tile_t.size().cols();
+          DLAF_ASSERT(taus.size() == k, taus.size(), k);
+
+          if (is_v0) {
+            cudaStream_t stream;
+            DLAF_CUBLAS_CALL(cublasGetStream(handle, &stream));
+
+            memory::MemoryView<T, Device::GPU> taus_d(taus.size());
+            DLAF_CUDA_CALL(cudaMemcpyAsync(taus_d(), taus.data(), to_sizet(taus.size()) * sizeof(T),
+                                           cudaMemcpyDefault, stream));
+
+            // manage computations with implicit 1 for the whole j loop
+            tfactorImplicit1(k, taus_d(), tile_v.ptr(), tile_v.ld(), tile_t.ptr(), tile_t.ld(), stream);
+          }
+
+          for (SizeType j = 0; j < k; ++j) {
+            const auto mtau = util::blasToCublasCast(-taus[j]);
+            const auto one = util::blasToCublasCast(T{1});
+            // this is the x0 element of the reflector j
+            const TileElementIndex x0{j, j};
+
+            const TileElementIndex t_start{0, x0.col()};
+
+            const SizeType first_element_in_tile = is_v0 ? x0.row() : 0;
+
+            // T(0:j, j) = -tau . V(j:, 0:j)* . V(j:, j)
+            // [j x 1] = [(n-j) x j]* . [(n-j) x 1]
+            TileElementSize va_size{tile_v.size().rows() - first_element_in_tile, x0.col()};
+            TileElementIndex va_start{first_element_in_tile, 0};
+            TileElementIndex vb_start{first_element_in_tile, x0.col()};
+
+            if (is_v0) {
+              // skip already managed computations with implicit 1
+              va_start = va_start + TileElementSize{1, 0};
+              vb_start = vb_start + TileElementSize{1, 0};
+              va_size = va_size - TileElementSize{1, 0};
+            }
+
+            if (!va_size.isEmpty()) {
+              cublas::Gemv<T>::call(handle, CUBLAS_OP_C, to_int(va_size.rows()), to_int(va_size.cols()),
+                                    &mtau, util::blasToCublasCast(tile_v.ptr(va_start)),
+                                    to_int(tile_v.ld()), util::blasToCublasCast(tile_v.ptr(vb_start)), 1,
+                                    &one, util::blasToCublasCast(tile_t.ptr(t_start)), 1);
+            }
+          }
+          return std::move(tile_t);
+        });
```

Comment:
What would it take for this to be merged with the MC version into a generic version? I guess your key requirement here is that you want these scheduled on the same stream sequentially? And the second thing that may make separating those cublas calls are the conditionals? Is there something else? It may not be possible, I'm just asking to understand if there's something in the sender world that would make it easier to merge (parts of) the implementations.

Reply:
I think the key is like above: the different signature of the task is the main reason.